README.md
# flutter_key_generator

Secure unique 128, 192, 256 bit key generator for Flutter iOS & Android.

## Author

Hayr Hotoca | [@1limxapp](https://twitter.com/1limxapp)
## License

MIT


CHANGELOG.md
## 0.0.1

* first version

## 0.0.2
* update package android name

## 0.0.3
* update github url

## 0.0.4
* add dartdoc comments

## 0.0.5
* change AES to ChaCha20 on android

## 0.0.6
* format dart code


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_key_generator_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_key_generator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_key_generator/flutter_key_generator.dart';
import 'package:flutter_key_generator/flutter_key_generator_platform_interface.dart';
import 'package:flutter_key_generator/flutter_key_generator_method_channel.dart';

class MockFlutterKeyGeneratorPlatform implements FlutterKeyGeneratorPlatform {
  @override
  Future<List<int>?> generateSymmetricKey(int size) => Future.value([42]);
}

void main() {
  final FlutterKeyGeneratorPlatform initialPlatform =
      FlutterKeyGeneratorPlatform.instance;

  test('$MethodChannelFlutterKeyGenerator is the default instance', () {
    expect(initialPlatform, isInstanceOf<MethodChannelFlutterKeyGenerator>());
  });

  test('generateSymmetricKey', () async {
    expect(await FlutterKeyGenerator.generateSymmetricKey(256), [42]);
  });
}


test/flutter_key_generator_method_channel_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_key_generator/flutter_key_generator_method_channel.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  MethodChannelFlutterKeyGenerator platform =
      MethodChannelFlutterKeyGenerator();
  const MethodChannel channel = MethodChannel('flutter_key_generator');

  setUp(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(
      channel,
      (MethodCall methodCall) async {
        return [42];
      },
    );
  });

  tearDown(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(channel, null);
  });

  test('generateSymmetricKey', () async {
    expect(await platform.generateSymmetricKey(256), [42]);
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_key_generator/flutter_key_generator.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  void genUniqueKey() async {
    try {
      final key = await FlutterKeyGenerator.generateSymmetricKey(256);
      print(key);
    } catch (e) {
      print(e);
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Running on:'),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: genUniqueKey,
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ), // This trailing comma makes auto-formatting nicer for build methods.
      ),
    );
  }
}


ios/Classes/FlutterKeyGeneratorPlugin.swift
import Flutter
import UIKit
import CryptoKit

public class FlutterKeyGeneratorPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "flutter_key_generator", binaryMessenger: registrar.messenger())
    let instance = FlutterKeyGeneratorPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      guard let args = call.arguments as? [String: Any] else {
          result(FlutterError(
              code: "CAUGHT_ERROR",
              message:"Invalid arguments",
              details: nil))
          return
      }

      do {
          switch call.method {
            case "generateSymmetricKey":
              try self.generateSymmetricKey(args: args, result: result)
            default:
              result(FlutterMethodNotImplemented)
            }
      } catch let error as NSError {
          result(FlutterError(
              code: "CAUGHT_ERROR",
              message:"\(error.domain), \(error.code), \(error.description)",
              details: nil))
      } catch {
          result(FlutterError(
              code: "CAUGHT_ERROR",
              message:"\(error)",
              details: nil))
      }
  }
    
    private func generateSymmetricKey(args: [String: Any], result: @escaping FlutterResult)  throws {
        if #available(iOS 13.0, *) {
            guard let size = args["size"] as? Int else {
                result(parameterError(name: "size"))
                return
            }

            var newKey = SymmetricKey.init(size: SymmetricKeySize.bits256)
            switch size {
              case 128 :
                newKey = SymmetricKey.init(size: SymmetricKeySize.bits128)
              case 192:
                newKey = SymmetricKey.init(size: SymmetricKeySize.bits192)
              case 256:
                newKey = SymmetricKey.init(size: SymmetricKeySize.bits256)
              default:
                result(FlutterError(code: "Key size error", message: "Key size should be 128, 192 or 256", details: nil))
                return
            }

            let key = newKey.withUnsafeBytes {
                return Data(Array($0))
            }
            
            result(FlutterStandardTypedData(bytes: key))
        }
        result(FlutterError(code: "UNSUPPORTED_ALGORITHM", message:nil, details: nil))
    }
    
    private func parameterError(name: String) -> FlutterError {
        return FlutterError(code: "INVALID_ARGUMENT", message: "Parameter '\(name)' is missing or invalid", details: nil)
    }
}


android/src/main/kotlin/com/hayr_hotoca/flutter_key_generator/FlutterKeyGeneratorPlugin.kt
package com.hayr_hotoca.flutter_key_generator

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey

/** FlutterKeyGeneratorPlugin */
class FlutterKeyGeneratorPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_key_generator")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "generateSymmetricKey") {
      generateSymmetricKey(call, result)
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  private fun generateSymmetricKey(call: MethodCall, result: Result) {
    try {
      val size = call.argument<Int>("size")
      if (size != 128 && size != 192 && size != 256) {
        return result.error(
        "Key size error",
        "Key size should be 128, 192 or 256.",
        null
        )
      }

      val keygen = KeyGenerator.getInstance("ChaCha20")
      keygen.init(size)
      val key: SecretKey = keygen.generateKey()

      result.success(key.encoded)
    } catch (e: Exception) {
      result.error("Generate symmetric key", e.localizedMessage, e)
    }
  }
}
